internal/ntobjectdir/enum_object_dir.go

```diff
@@ -0,0 +1,111 @@
+package ntobjectdir
+
+import (
+	"path/filepath"
+	"syscall"
+	"unicode/utf16"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/kernel32"
+	"github.com/Microsoft/hcsshim/internal/ntconstants"
+)
+
+//go:generate go run ..\..\mksyscall_windows.go -output zsyscall_windows.go enum_object_dir.go
+
+//sys ntOpenDirectoryObject(handle *uintptr, accessMask uint32, oa *ntconstants.ObjectAttributes) (status uint32) = ntdll.NtOpenDirectoryObject
+//sys ntQueryDirectoryObject(handle uintptr, buffer *byte, length uint32, singleEntry bool, restartScan bool, context *uint32, returnLength *uint32)(status uint32) = ntdll.NtQueryDirectoryObject
+
+const globalNTPath = "\\Global??"
+const bufferSize = 1024
+
+// ntWidePath converts a golang string to a wide string
+func ntWidePath(path string) ([]uint16, error) {
+	path = filepath.Clean(path)
+	fspath := filepath.FromSlash(path)
+	path16 := utf16.Encode(([]rune)(fspath))
+	if len(path16) > 32767 {
+		return nil, syscall.ENAMETOOLONG
+	}
+	return path16, nil
+}
+
+// EnumerateNTGlobalObjectDirectory queries `globalNTPath` for all entires and
+// returns the resulting entry names as a string slice.
+func EnumerateNTGlobalObjectDirectory() ([]string, error) {
+	var (
+		handle uintptr
+		oa     ntconstants.ObjectAttributes
+
+		context      uint32
+		returnLength uint32
+		buffer       [bufferSize]byte
+		result       []string
+	)
+
+	globalNTPath16, err := ntWidePath(globalNTPath)
```

Comment:
If we're doing all the work to build a function that can enumerate the object tree, it seems useful to make it take a path argument instead of hardcoding the path. That way we can easily re-use it elsewhere.

Reply:
Fair enough. Done :)